### PR TITLE
remove nemo.collection.nlp imports from nemo2

### PR DESCRIPTION
**Pull request by dimapihtar**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/llm/gpt/model/base.py [hunk 1]
```diff
@@ -442,6 +447,286 @@ def validation_loss_reduction(self) -> MaskedTokenLossReduction:
...
+                attention_mask=attention_mask,
+                encoder_output=encoder_output,
+                enc_dec_attn_mask=enc_dec_attn_mask,
+                inference_params=inference_params,
+                is_first_microbatch=is_first_microbatch,
+                checkpoint_core_attention=checkpoint_core_attention,
+            )
+
+
+class TETransformerLayerAutocast(AutocastTransformerLayer, BaseTransformerLayer):
```
Comment (ericharper) [inline nemo/collections/llm/gpt/model/base.py:576]: Can we get this from mcore directly?

File: nemo/collections/llm/quantization/utils.py [hunk 2]
```diff
@@ -52,11 +53,9 @@ def get_modelopt_decoder_type(model: llm.GPTModel) -> str:
 def quantizable_model_config(model_cfg: llm.GPTConfig) -> llm.GPTConfig:
     """Modify model config for TensorRT-Model-Optimizer quantization"""
 
-    from nemo.collections.nlp.models.language_modeling.megatron.gpt_layer_modelopt_spec import (
-        get_gpt_layer_modelopt_spec,
+    model_cfg.transformer_layer_spec = get_gpt_layer_modelopt_spec(
+        num_experts=model_cfg.num_moe_experts, remap_te_layernorm=True
```
Comment (janekl) [inline nemo/collections/llm/quantization/utils.py:57]: ✅

File: nemo/collections/llm/quantization/utils.py [hunk 3]
```diff
@@ -52,11 +53,9 @@ def get_modelopt_decoder_type(model: llm.GPTModel) -> str:
...
     )
-
-    model_cfg.transformer_layer_spec = get_gpt_layer_modelopt_spec(num_experts=model_cfg.num_moe_experts)
```
Comment (janekl) [inline nemo/collections/llm/quantization/utils.py:59]: LGTM as for changes to `nemo/collections/llm/quantization` ✅

File: nemo/collections/llm/bert/loss.py [hunk 4]
```diff
@@ -241,3 +238,16 @@ def sentence_order_prediction_loss(tensor: Tensor, sentence_order: Tensor):
...
+    """Reduce a tensor of losses across all GPUs."""
+    from megatron.core import parallel_state
+
+    averaged_losses = torch.cat([loss.clone().detach().view(1) for loss in losses])
+    torch.distributed.all_reduce(averaged_losses, group=parallel_state.get_data_parallel_group())
+    averaged_losses = averaged_losses / torch.distributed.get_world_size(
+        group=parallel_state.get_data_parallel_group()
+    )
+
+    return averaged_losses
```
Comment (ananthsub) [inline nemo/collections/llm/bert/loss.py:253]: instead of copy/pasting, does it make sense to move this from `nemo.collections.nlp.modules.common.megatron.utils` to somewhere like here? https://github.com/NVIDIA/NeMo/blob/main/nemo/utils/distributed.py

and have nemo1 and nemo2 use the same function then?

File: nemo/collections/llm/gpt/model/gemma2.py [hunk 5]
```diff
@@ -264,6 +300,271 @@ def config(self) -> "Gemma2Config":
...
+    """
+
+    def __init__(
+        self,
+        config: TransformerConfig,
+        layer_number: int,
+        attn_mask_type: AttnMaskType,
+        attention_type: str,
+        attention_dropout: float = None,
+        **kwargs,
```
Comment (ananthsub) [inline nemo/collections/llm/gpt/model/gemma2.py:326]: where are the kwargs used?

File: nemo/collections/llm/gpt/model/gemma2.py [hunk 6]
```diff
@@ -264,6 +300,271 @@ def config(self) -> "Gemma2Config":
...
+
+    def forward(
+        self,
+        query: Tensor,
+        key: Tensor,
+        value: Tensor,
+        attention_mask: Tensor,
+        attn_mask_type: AttnMaskType = None,
+        packed_seq_params: PackedSeqParams = None,
+        **kwargs,
```
Comment (ananthsub) [inline nemo/collections/llm/gpt/model/gemma2.py:386]: same question here

File: nemo/collections/llm/utils.py [hunk 7]
```diff
@@ -75,3 +77,16 @@ def noop_decorator(func: T) -> T:
...
+
+def torch_dtype_from_precision(precision: Union[int, str]) -> torch.dtype:
+    """Mapping from PTL precision types to corresponding PyTorch parameter datatype."""
+
+    if precision in ['bf16', 'bf16-mixed']:
+        return torch.bfloat16
+    elif precision in [16, '16', '16-mixed']:
+        return torch.float16
+    elif precision in [32, '32', '32-true']:
+        return torch.float32
```
Comment (ananthsub) [inline nemo/collections/llm/utils.py:90]:
```suggestion
    if precision in ('bf16', 'bf16-mixed'):
        return torch.bfloat16
    elif precision in (16, '16', '16-mixed'):
        return torch.float16
    elif precision in (32, '32', '32-true'):
        return torch.float32
```

File: nemo/collections/llm/bert/loss.py (hunk 4 above)
Comment (dimapihtar) [reply nemo/collections/llm/bert/loss.py:253]: > instead of copy/pasting, does it make sense to move this from `nemo.collections.nlp.modules.common.megatron.utils` to somewhere like here? https://github.com/NVIDIA/NeMo/blob/main/nemo/utils/distributed.py
> 
> and have nemo1 and nemo2 use the same function then?

@ananthsub looks like this function is only used for bert. Also, we are going to deprecate nemo1 soon so we don't need to have this function for nemo1 in future

File: nemo/collections/llm/gpt/model/gemma2.py (hunk 5 above)
Comment (dimapihtar) [reply nemo/collections/llm/gpt/model/gemma2.py:326]: > where are the kwargs used?

@ananthsub not sure, but there was a fix some time ago which added kwargs so I think we want to keep them here:
https://github.com/NVIDIA/NeMo/pull/11792

File: nemo/collections/llm/gpt/model/gemma2.py (hunk 6 above)
Comment (dimapihtar) [reply nemo/collections/llm/gpt/model/gemma2.py:386]: I think here is the same situation as with init method

File: nemo/collections/llm/gpt/data/helpers.cpp
Comment (ericharper) [inline nemo/collections/llm/gpt/data/helpers.cpp:1]: Can we remove this and just get it from megatron?

File: nemo/collections/llm/gpt/data/core.py [hunk 8]
```diff
@@ -77,25 +109,1334 @@ def create_sft_dataset(
...
+        try:
+            record = json.loads(text)
+        except Exception as e:
+            logging.error(f"Exception: {e}")
+            logging.error(f"datapoint: {text}")
+            raise e
+        return record
+
+
+class OnlineSampleMapping:
```
Comment (cuichenx) [inline nemo/collections/llm/gpt/data/core.py:1267]: I think it makes more sense if TextMemMapDataset, JSONLMemMapDataset and OnlineSampleMapping are moved to a different file (maybe utils.py or maybe another file). Also, it might make sense to add a _ prefix before them to indicate that they're not to be instantiated by users

File: nemo/collections/llm/gpt/data/utils.py [hunk 9]
```diff
@@ -0,0 +1,643 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
```
Comment (cuichenx) [inline nemo/collections/llm/gpt/data/utils.py:1]:
```suggestion
# Copyright (c) 2025, NVIDIA CORPORATION.  All rights reserved.
```

**Review by cuichenx — COMMENTED** [review]
LGTM on data and peft front with two very minor points..

File: nemo/collections/llm/gpt/data/helpers.cpp
Comment (dimapihtar) [reply nemo/collections/llm/gpt/data/helpers.cpp:1]: removed and imported helpers from mcore. ci pipeline doesn't show any issue.

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!

**Review by ericharper — CHANGES_REQUESTED** [review]
Blocking until we code freeze.
